### DISCUSSION TITLE
I would like to inquire about how to build a Windows binary file (exe).

### DISCUSSION THREAD

**Discussion by TxcA** [milvus-io/milvus#37676] [Q&A and General discussion] [answered]
I am using Alibaba Cloud's ECS with Windows Server 2019. Since ECS is already a virtual machine, I am unable to install Docker (or rather, I cannot virtualize a Linux environment). I came across the project '[matrixji/milvus](https://github.com/matrixji/milvus)', but unfortunately, it hasn't been maintained for a long time.
Therefore, I would like to try compiling the exe file myself. Thank you for your response.

**Comment by yhmo** [comment]
https://github.com/milvus-io/milvus/blob/v2.4.15/DEVELOPMENT.md#software-requirements

A list of verified OS types where Milvus can successfully build and run:
- Debian/Ubuntu
- Amazon Linux
- MacOS (x86_64)
- MacOS (Apple Silicon)

**Reply by TxcA** [reply]
> https://github.com/milvus-io/milvus/blob/v2.4.15/DEVELOPMENT.md#software-requirements
> 
> A list of verified OS types where Milvus can successfully build and run:
> 
> * Debian/Ubuntu
> * Amazon Linux
> * MacOS (x86_64)
> * MacOS (Apple Silicon)

Thank you. The reason there is no update for `matrixji/milvus` 2.4 is that it can no longer package into an Windows(exe) after version 2.4?

**Reply by yhmo** [reply]
I think the reason is, with the milvus source code requires a much higher soft environment, it is difficult to build milvus on windows.

**Reply by TxcA** [reply]
Thanks, I'll think of other solutions.

**Comment by TxcA** [comment]
@matrixji Hello, could you share the build method for Windows?

**Reply by xiaofan-luan** [reply]
I don't think it's doable.
A lot of the dependencies has been changed really quick for the last couple of years.

We will have a document about how to run milvus with WSL or Docker on windows and hopefully that could help

**Reply by matrixji** [reply, marked as answer]
Hi, @TxcA
Porting milvus to Windows is not an easy job, I started to do porting from milvus 2.0, and the last successful version is 2.2.16 you can find [patch here](https://github.com/milvus-io/milvus-lite/blob/2.2/milvus_binary/patches/milvus-v2.2.16/0001-patch-for-milvus-lite.patch), from milvus 2.3, I could not handle it, for it's fast-growing code. At the same time, the application of vector databases has undergone many changes. I started using milvus from 0.10, at that moment, the main application area of vector databases was image similarity search. We hope for high performance, we hope high recall, and we hope can debug/test on a local computer easily. Recently, especially with the rise of large language models, the usage scenarios of vector databases have undergone significant changes. How to deploy is no longer important, but the need for more convenient use of Milvus has always existed, which is also the reason why Milvus Lite continues to be active. But personally, if I want to use a vector database now, I am more inclined to use SaaS for quick validation solutions, and when it comes to actual implementation, I will be more inclined to align with the existing infrastructure architecture (obviously there won't be Windows here)

**Reply by matrixji** [reply]
For SaaS Zilliz provides free small instances, it's a good solution for some startup PoC projects.

**Reply by TxcA** [reply]
> I don't think it's doable. A lot of the dependencies has been changed really quick for the last couple of years.
> 
> We will have a document about how to run milvus with WSL or Docker on windows and hopefully that could help

Thank you very much for your answer. Since I am using an ECS, secondary virtualization is not possible, so WSL is unavailable.

**Reply by TxcA** [reply]
> 2.2.16

Thank you for your response. I'll try version 2.2.16. We're a small-scale operation. Thank you very much!

**Reply by xiaofan-luan** [reply]
> > 2.2.16
> 
> Thank you for your response. I'll try version 2.2.16. We're a small-scale operation. Thank you very much!

I don't recommend to do so. 

you can still install a docker in your ecs and should not be very hard.

@LoveEachDay is working on a document on how to deploy milvus in windows with Docker or K8s env.

Since you already use ecs, why not buy a linux ecs instead?

**Reply by TxcA** [reply]
> > > 2.2.16
> > 
> > 
> > Thank you for your response. I'll try version 2.2.16. We're a small-scale operation. Thank you very much!
> 
> I don't recommend to do so.
> 
> you can still install a docker in your ecs and should not be very hard.
> 
> @LoveEachDay is working on a document on how to deploy milvus in windows with Docker or K8s env.
> 
> Since you already use ecs, why not buy a linux ecs instead?

---

Docker deployment and Kubernetes testing have been successful on my local machine. However, since ECS does not support nested virtualization, Docker can only run in a Windows environment. As our main project is based on C#, we need to use Windows Server.

We initially used OpenSearch with the k-NN plugin. However, since our dataset is relatively small (around 500,000 entries) and we only perform image similarity searches, OpenSearch felt overly cumbersome for our needs. After nearly two years of operation, we had to consider switching.

After comparing several options, we chose Milvus. Unfortunately, it doesn’t offer good support for Windows. We have since migrated to Qdrant (though it’s probably not appropriate to mention other projects here, I must say Qdrant is very Windows-friendly), and it’s been running smoothly so far.

Thank you very much for your response.
